src/VectorBundle.jl

```diff
@@ -0,0 +1,221 @@
+for manifold `M` on given arguments (passed at a tuple).
+"""
+function similar_result_type(M::Manifold, VS::VectorSpaceType, f, args::NTuple{N,Any}) where N
+    T = typeof(reduce(+, one(eltype(eti)) for eti ∈ args))
+    return T
+end
+
+"""
+    similar_result(M::Manifold, VS::VectorSpaceType, f, x...)
+
+Allocates an array for the result of function `f` that is
+an element of the vector space of type `VS` on manifold `M`
+and arguments `x...` for implementing the non-modifying operation
+using the modifying operation.
+"""
+function similar_result(M::Manifold, VS::VectorSpaceType, f, x...)
+    T = similar_result_type(M, VS, f, x)
+    return similar(x[1], T)
+end
+
+
+"""
+    manifold_dimension(M::Manifold, VS::VectorSpaceType)
+
+Dimension of the vector space of type `VS` over manifold `M`.
+"""
+function manifold_dimension(M::Manifold, VS::VectorSpaceType)
+    error("manifold_dimension not implemented for manifold $(typeof(M)) and vector space of type $(typeof(VS)).")
+end
+
+manifold_dimension(M::Manifold, ::TCoTSpaceType) = manifold_dimension(M)
+
+function representation_size(M::Manifold, VS::TCoTSpaceType)
+    representation_size(M)
+end
+
+"""
+    zero_vector!(M::Manifold, VS::VectorSpaceType, v, x)
+
+Save the zero vector from the vector space of type `VS` at point `x`
+from manifold `M` to `v`.
+"""
+function zero_vector!(M::Manifold, VS::VectorSpaceType, v, x)
+    error("zero_vector! not implemented for manifold $(typeof(M)), vector space of type $(typeof(VS)), vector of type $(typeof(v)) and point of type $(typeof(x)).")
+end
+
+function zero_vector!(M::Manifold, VS::TCoTSpaceType, v, x)
+    zero_tangent_vector!(M, v, x)
+    return v
+end
+
+"""
+    zero_vector(M::Manifold, VS::VectorSpaceType, x)
+
+Compute the zero vector from the vector space of type `VS` at point `x`
+from manifold `M`.
+"""
+function zero_vector(M::Manifold, VS::VectorSpaceType, x)
+    v = similar_result(M, VS, zero_vector, x)
+    zero_vector!(M, VS, v, x)
+    return v
+end
+
+"""
+    project_vector!(M::Manifold, VS::VectorSpaceType, v, x, w)
+
+Project vector `w` from the vector space of type `VS` at point `x`
+and save the result to `v`.
+"""
+function project_vector!(M::Manifold, VS::VectorSpaceType, v, x, w)
+    error("project_vector! not implemented for vector space manifold $(typeof(M)), vector space of type $(typeof(VS)), output vector of type $(typeof(v)) and input vector at point $(typeof(x)) with type of w $(typeof(w)).")
+end
+
+"""
+    inner(M::Manifold, VS::VectorSpaceType, x, v, w)
+
+Inner product of vectors `v` and `w` from the vector space of type `VS`
+at point `x` from manifold `M`.
+"""
+function inner(M::Manifold, VS::VectorSpaceType, x, v, w)
+    error("inner not defined for manifold of type $(typeof(M)), vector space of type $(typeof(VS)), point of type $(typeof(x)) and vectors of types $(typeof(v)) and $(typeof(w)).")
+end
+
+function inner(M::Manifold, VS::TCoTSpaceType, x, v, w)
```

Comment:
Is this a suitable default for a `CotangentSpace`? Doesn't the cotangent vector `v` need to be converted to a tangent vector before defining the inner product? Actually, maybe we need a general method for interconverting the two. It could just error by default but for `MetricManifold` use the metric directly.

Reply:
This is a very good question. I was thinking that covectors will have the same representation as their duals and thus computing their inner product would be easy but this assumption is indeed wrong. We probably need a method like `dual!(M::Manifold, VS::VectorSpaceType, v, x, w)` that saves to `v` the dual representation of vector `w` from vector space of type `VS` at `x`.

Reply:
For me inner on tangent space introduces an inner product on the cotangent space by the musical isomorphisms. While I am pro a conversion between tangents and cotangents, `dual` for me might also refer to a dual pairing between tangent and cotangent space, so we should discuss this representation-change-functions name maybe.

Reply:
> For me inner on tangent space introduces an inner product on the cotangent space by the musical isomorphisms.

This is also how I interpreted it. If my math is correct, a vector is obtained from a covector by multiplication with the cometric, i.e. `v^j = g^{ji} v_i`. The inner product for vectors is `g_{ij} v^i w^j`. Therefore, the inner product in terms of covectors is simply `g^{ij} v_i w_j`.

We could by default raise an error message for covector inner products, and individual manifolds are free to implement them. For `MetricManifold`, we could compute the inner product by default using `inverse_local_metric `, much in the same way as the `inner` on the tangent space does with `local_metric` for these manifolds. This has the benefit of being consistent with what we already do.

Reply:
Wouldn't it be better to implement the default covector inner product using the sharp isomorphism? And just make the musical isomorphisms raise an error by default?

Reply:
Yes, good idea. I think what I proposed may be a better default for `MetricManifold`.